lib: add fileContents function

ericsagnes opened the pull request:
###### Motivation for this change

The pattern `removeSuffix "\n" (builtins.readFile file)` is very common to read version numbers or passwords from a single line file.

Having the functionality in lib would save some typing and [search](https://www.mail-archive.com/nix-dev@lists.science.uu.nl/msg20298.html).
###### Things done
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

ericsagnes committed:
lib: add fileContents function

domenkozar commented on lib/strings.nix:
```diff
@@ -479,4 +479,8 @@ rec {
       absolutePaths;
+
+  /* Read the contents of a file removing the trailing \n
+  */
```
Lacks an example usage as other functions.

zimbatm commented:
Thanks, I definitely had the need to do that in the past.

If you could also refactor the existing code to take advantage of that function, the motivation of the PR would be much stronger.

domenkozar commented:
We should use this to extract `.version` at least :)

It would make my life easier, since a lot of times I broke nixpkgs when bumping NixOS verison and introduced a newline.

ericsagnes committed:
lib: refactor commitIdFromGitRepo with fileContents

ericsagnes committed:
lib: refactor nixpkgsVersion with fileContents

ericsagnes committed:
release.nix: refactor with fileContents

ericsagnes committed:
version module: refactor with fileContents

ericsagnes commented:
Added an example and refactored all the functions that I found.
Please let me know if I forgot something.

zimbatm commented:
Nice!
I found another one at pkgs/top-level/make-tarball.nix:18

ericsagnes committed:
make tarball: refactor with fileContents

ericsagnes commented:
> I found another one at pkgs/top-level/make-tarball.nix:18

Nice catch, I added a commit for that and rebased on master because there was a conflict in the version module.

zimbatm commented:
Thanks!

edolstra commented:
The name `fileContents` seems bad to me, since it in no way suggests that it strips a trailing newline. Maybe it should be something like `chompFile` (following Perl's `chomp` function).

zimbatm commented:
Let's copy Haskell. Isn't there a function in Haskell that does that?
By the way, just reading the first line of a file would be enough for all the use-cases.
